0002-chunking.md

```diff
@@ -0,0 +1,271 @@
+```
+for some shared public generators `g_0`, ..., `g_n`. These commitments are integral to the kimchi proof system, where we use the values of the polynomials to represent the list of values that appear in the proof. For example, the list
+```text
+[b_0, b_1, ..., b_{n-1}]
+```
+is represented by the polynomial `B` over the domain generated by `w` as
+```text
+B(w^0) = b_0
+B(w^1) = b_1
+...
+B(w^{n-1}) = b_{n-1}
+```
+The polynomial `B` of degree `n-1` is uniquely determined by these equations, and the domain generator satisfies `w^n = 1`, `w^i != 1` for `0 < i < n`.
+
+In order to support circuits of a given length, we need to represent the values at each step (or 'gate') in the proof, and thus to represent them as a polynomial of the degree. However, kimchi+pickles as proposed for the Berkeley hard fork are limited to `2^16` generators for performance.
+
+This RFC proposes splitting the polynomial into chunks when we build the commitment. For example, we could represent the polynomial
+```text
+f(x) = a_0 x^0 + a_1 x^1 + ... + a_{4*n-1} x^{4*n-1}
+```
+by the commitments
+```text
+C_0 = g_0^{a_0} + g_1^{a_1} + ... + g_n^{a_{n-1}}
+C_1 = g_0^{a_n} + g_1^{a_{n+1}} + ... + g_n^{a_{2n-1}}
+C_2 = g_0^{a_{2n}} + g_1^{a_{2n+1}} + ... + g_n^{a_{3n-1}}
+C_3 = g_0^{a_{3n}} + g_1^{a_{3n+1}} + ... + g_n^{a_{4n-1}}
+```
+which is equivalent to committing to the 4 'chunked' polynomials
+```text
+f_0(x) = a_0 x^0 + a_1 x^1 + ... + a_{n-1} x^{n-1}
+f_1(x) = a_n x^0 + a_{n+1} x^1 + ... + a_{2n-1} x^{n-1}
+f_2(x) = a_{2n} x^0 + a_{2n+1} x^1 + ... + a_{3n-1} x^{n-1}
+f_3(x) = a_{3n} x^0 + a_{3n+1} x^1 + ... + a_{4n-1} x^{n-1}
+```
+that we can recompose to form `f`
+```text
+f(x) = f_0(x) + x^n * f_1(x) + x^{2n} * f_2(x) + x^{3n} * f_3(x)
+```
+
+When we check the constraint polynomial at a ['random'](https://en.wikipedia.org/wiki/Fiat%E2%80%93Shamir_heuristic) point `z` later in the proof, we can build a commitment that matches that point by computing
+```text
+C = C_0 + C_1^{z^n} + C_2^{z^{2n}} + C_3^{z^{3n}}
+```
+while also checking that the prover's claimed evaluations are consistent with the commitments:
+```text
+[( f_0(z), C_0 ), ( f_1(z), C_1 ), ( f_2(z), C_2 ), ( f_3(z), C_3 )]
+```
+Finally, we can compute the value `f(z)` by proxy:
+```text
+f(z) = f_0(z) + z^n f_1(z) + z^{2n} f_2(z) + z^{3n} f_3(z)
+```
+
+Thus, this chunking approach provides all of the features that we need to support the kimchi proof system over a larger domain size than the current Mina limit.
+
+### Retaining zero-knowledge
+
+Once we have `c > 1` chunks, every column will come with `2 * c` evaluations, 1 for each chunk at each of `z` and `z * w`. In order to retain zero-knowledge, we need to ensure that we randomize more columns than there are evaluations. Thus, we must randomize at least `2 * c + 1` values at the end of each column. The helpers for adding zero-knowledge to columns should be updated to take the number of zk rows that are required as an argument.
```

Comment:
```suggestion
After we have `c > 1` chunks, every column will come with `2 * c` evaluations, 1 for each chunk at each of `z` and `z * w`. In order to retain zero-knowledge, we need to ensure that we randomize more columns than there are evaluations. Thus, we must randomize at least `2 * c + 1` values at the end of each column. The helpers for adding zero-knowledge to columns must be updated to take the number of zk rows that are required as an argument.
```

Reply:
if we mean "must" let's be sure to use must, otherwise I am reading "should" as a suggestion